src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests4.cs

```diff
@@ -1489,6 +1489,211 @@ static int M1(bool? b1, bool? b2)
+                Diagnostic(ErrorCode.WRN_SwitchExpressionNotExhaustive, "switch").WithLocation(9, 19)
+                );
+            CompileAndVerify(compilation, expectedOutput: "MatchFailureException()");
+        }
+
+        [Fact]
+        public void UnmatchedInput_03()
+        {
+            var source =
+@"using System;
+public class C
+{
+    static void Main()
+    {
+        var t = (1, 2);
+        try
+        {
+            _ = t switch { (3, 4) => 1 };
+        }
+        catch (MatchFailureException ex)
+        {
+            Console.WriteLine($""{ex.GetType().Name}({ex.UnmatchedValue})"");
+        }
+        catch (Exception ex)
+        {
+            Console.WriteLine(ex.GetType().Name);
+        }
+    }
+}
+namespace System
+{
+    public class MatchFailureException : InvalidOperationException
+    {
+        public MatchFailureException() {}
+        public MatchFailureException(object unmatchedValue) => UnmatchedValue = unmatchedValue;
+        public object UnmatchedValue { get; }
+    }
+}
+";
+            var compilation = CreatePatternCompilation(source);
+            compilation.VerifyDiagnostics(
+                // (9,19): warning CS8509: The switch expression does not handle all possible inputs (it is not exhaustive).
+                //             _ = t switch { (3, 4) => 1 };
+                Diagnostic(ErrorCode.WRN_SwitchExpressionNotExhaustive, "switch").WithLocation(9, 19)
+                );
+            CompileAndVerify(compilation, expectedOutput: "MatchFailureException((1, 2))");
+        }
+
+        [Fact]
+        public void UnmatchedInput_04()
+        {
+            var source =
+@"using System;
+public class C
+{
+    static void Main()
+    {
+        try
+        {
+            _ = (1, 2) switch { (3, 4) => 1 };
+        }
+        catch (MatchFailureException ex)
+        {
+            Console.WriteLine($""{ex.GetType().Name}({ex.UnmatchedValue})"");
+        }
+        catch (Exception ex)
+        {
+            Console.WriteLine(ex.GetType().Name);
+        }
+    }
+}
+namespace System
+{
+    public class MatchFailureException : InvalidOperationException
+    {
+        public MatchFailureException() {}
+        public MatchFailureException(object unmatchedValue) => UnmatchedValue = unmatchedValue;
+        public object UnmatchedValue { get; }
+    }
+}
+";
+            var compilation = CreatePatternCompilation(source);
+            compilation.VerifyDiagnostics(
+                // (8,24): warning CS8509: The switch expression does not handle all possible inputs (it is not exhaustive).
+                //             _ = (1, 2) switch { (3, 4) => 1 };
+                Diagnostic(ErrorCode.WRN_SwitchExpressionNotExhaustive, "switch").WithLocation(8, 24)
+                );
+            CompileAndVerify(compilation, expectedOutput: "MatchFailureException()");
+        }
+
+        [Fact]
+        public void UnmatchedInput_05()
```

Comment:
>UnmatchedInput_05 [](start = 20, length = 17)

Please add a test for a scenario when neither InvalidOperationException, nor MatchFailureException are available at compile time.